src/helpers/author-archive-helper.php

```diff
@@ -23,6 +41,53 @@ public function get_author_archive_post_types() {
 		return \apply_filters( 'wpseo_author_archive_post_types', [ 'post' ] );
 	}
 
+	/**
+	 * Checks whether the user with the given ID has publicly viewable posts.
+	 *
+	 * **Note**: It uses the WordPress tables to determine the number of posts,
+	 * not the indexables table.
+	 *
+	 * @param string $user_id The user ID.
+	 *
+	 * @return bool Whether the author has public posts, according to the WordPress tables.
+	 */
+	public function author_has_public_posts_wp( $user_id ) {
+		$cache_key       = 'author_has_a_public_post_wp_' . $user_id;
+		$has_public_post = \wp_cache_get( $cache_key );
+
+		if ( $has_public_post === false ) {
+			global $wpdb;
+
+			$post_types        = $this->get_author_archive_post_types();
+			$public_post_stati = \array_values( \array_filter( \get_post_stati(), 'is_post_status_viewable' ) );
+
+			$post_type_placeholders         = \implode( ', ', \array_fill( 0, \count( $post_types ), '%s' ) );
+			$public_post_stati_placeholders = \implode( ', ', \array_fill( 0, \count( $public_post_stati ), '%s' ) );
+
+			$sql = "
+				SELECT ID 
+				FROM $wpdb->posts 
+				WHERE post_author = %d 
+				  AND post_type IN ( $post_type_placeholders ) 
+				  AND post_status IN ( $public_post_stati_placeholders )
+				LIMIT 1";
+
+			$replacements = \array_merge( [ $user_id ], $post_types, $public_post_stati );
+
+			// phpcs:disable WordPress.DB -- Query is prepared with placeholders above, query is cached.
+			$single_public_post = $wpdb->query( $wpdb->prepare( $sql, $replacements ) );
```

Comment:
**Note for reviewers:** I'm using a separate query here since we do not know whether the posts in the indexable table are up to date when building the author.

More specifically: when we build a post indexable we also build its author indexable if it does not exist yet. This leads to a catch-22: we can only check whether an author has any posts (using the indexable table) when the author indexable already exists and has an ID.

As a solution we query the WordPress posts table instead.